### PR TITLE
C#: Add flow summary for 'Task.ConfigureAwait()'

### PR DESCRIPTION
**Pull request by tamasvajk**

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/dataflow/LibraryTypeDataFlow.qll [hunk 1]
```diff
@@ -1843,6 +1843,45 @@ class SystemThreadingTasksTaskTFlow extends LibraryTypeDataFlow, SystemThreading
...
+    sink = TCallableFlowSinkReturn() and
+    sinkAp =
+      AccessPath::cons(any(FieldContent fc |
+          fc.getField() =
+            any(SystemRuntimeCompilerServicesConfiguredTaskAwaitableTStruct t)
+                .getUnderlyingAwaiterField()
+        ),
+        AccessPath::field(any(SystemRuntimeCompilerServicesConfiguredTaskAwaitableTConfiguredTaskAwaiterStruct s
+          ).getUnderlyingTaskField()))
+  }
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/dataflow/LibraryTypeDataFlow.qll:1864]: @hvitved This is missing from the test expected file.
When I try to debug this on a `codeql database create` DB with the following:
```
private predicate test(
  SystemRuntimeCompilerServicesConfiguredTaskAwaitableTConfiguredTaskAwaiterStruct a, Field f
) {
  a.getAField() = f
}
```
I only get two fields: `_dummy` and `_dummyPrimitive`, and therefore I get no fields in `.getUnderlyingTaskField()`. It looks to me that Roslyn reports these field names.

* What is the difference between the test and a `codeql database create` extraction? 
* How do I extract the correct field names with `codeql database create`? 
* Why doesn't this `AccessPath::cons` work?

**Review by tamasvajk — COMMENTED** [review]
Let's suppose the `AccessPath::cons` is working. How would I verify that the added flow summaries are working for the following two cases?
```
var i = new Task<int>(() => 42).ConfigureAwait(false).GetAwaiter().GetResult();
var j = await new Task<int>(() => 42).ConfigureAwait(false);
```
This question is actually not specific to `.ConfigureAwait()`. Which checks rely on flow summaries on `.GetAwaiter().GetResult()`?

File: csharp/ql/src/semmle/code/csharp/dataflow/LibraryTypeDataFlow.qll [hunk 2]
```diff
@@ -1876,11 +1884,11 @@ class SystemRuntimeCompilerServicesConfiguredTaskAwaitableTFlow extends LibraryT
     // var result = awaiter.GetResult();
     c = this.getGetAwaiterMethod() and
     source = TCallableFlowSourceQualifier() and
-    sourceAp = AccessPath::empty() and
-    sink = TCallableFlowSinkReturn() and
-    sinkAp =
+    sourceAp =
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/dataflow/LibraryTypeDataFlow.qll:1887]: hmm, I don't know how I messed this up with the explicitly added comment above it. Thanks.

**Comment by tamasvajk** [comment]
@hvitved The only change I did since your fix is that I've squashed the first two commits.

File: csharp/ql/src/semmle/code/csharp/frameworks/system/runtime/CompilerServices.qll [hunk 3]
```diff
@@ -28,3 +28,38 @@ class SystemRuntimeCompilerServicesTaskAwaiterStruct extends SystemRuntimeCompil
...
+  /** Gets the `GetAwaiter` method. */
+  Method getGetAwaiterMethod() { result = this.getAMethod("GetAwaiter") }
+
+  /** Gets the field that stores the underlying awaiter. */
+  Field getUnderlyingAwaiterField() {
+    result = this.getAField() and result.hasName("m_configuredTaskAwaiter")
+  }
+}
+
+/** An unbound generic struct in the `System.Runtime.CompilerServices` namespace. */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/frameworks/system/runtime/CompilerServices.qll:47]: This comment is not correct. I think this class is so special that there is no need to have it, instead inline the logic where it is used below.

**Review by hvitved — CHANGES_REQUESTED** [review]

**Review by hvitved — APPROVED** [review]

**Pull request merged**